module/VuFind/src/VuFind/RateLimiter/RateLimiterManager.php

```diff
@@ -0,0 +1,299 @@
+<?php
+
+/**
+ * Rate limiter manager.
+ *
+ * PHP version 8
+ *
+ * Copyright (C) The National Library of Finland 2024.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Cache
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+namespace VuFind\RateLimiter;
+
+use Closure;
+use Laminas\EventManager\EventInterface;
+use Laminas\Log\LoggerAwareInterface;
+use Laminas\Mvc\MvcEvent;
+use VuFind\I18n\Translator\TranslatorAwareInterface;
+use VuFind\I18n\Translator\TranslatorAwareTrait;
+use VuFind\Log\LoggerAwareTrait;
+use VuFind\Net\IpAddressUtils;
+
+use function in_array;
+use function is_bool;
+
+/**
+ * Rate limiter manager.
+ *
+ * @category VuFind
+ * @package  Cache
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+class RateLimiterManager implements LoggerAwareInterface, TranslatorAwareInterface
+{
+    use LoggerAwareTrait;
+    use TranslatorAwareTrait;
+
+    /**
+     * Current event description for logging
+     *
+     * @var string
+     */
+    protected $eventDesc = '??';
+
+    /**
+     * Client ID for logging
+     *
+     * @var string
+     */
+    protected $clientId;
+
+    /**
+     * Constructor
+     *
+     * @param array          $config                     Rate limiter configuration
+     * @param string         $clientIp                   Client's IP address
+     * @param ?string        $userId                     User ID or null if not logged in
+     * @param Closure        $rateLimiterFactoryCallback Rate limiter factory callback
+     * @param IpAddressUtils $ipUtils                    IP address utilities
+     */
+    public function __construct(
+        protected array $config,
+        protected string $clientIp,
+        protected ?string $userId,
+        protected Closure $rateLimiterFactoryCallback,
+        protected IpAddressUtils $ipUtils
+    ) {
+        $this->clientId = "ip:$clientIp";
```

Comment:
Since client ID handling has moved to the factory, and this value is only being used for debug purposes, might it be worth renaming this property to something like clientDebugDetails, and always include IP address for more complete information?

Reply:
Changed to clientLogDetails since it's used for non-debug logging too.